py/desispec/preproc.py

```diff
@@ -0,0 +1,275 @@
+        rdnoise *= gain
+        kk = _parse_sec_keyword(header['CCDSEC'+amp])
+        readnoise[kk] = rdnoise
+        
+        header['OVERSCN'+amp] = overscan
+        header['OBSRDN'+amp] = rdnoise
+        
+        #- Warn/error if measured readnoise is very different from expected
+        if 'RDNOISE'+amp in header:
+            expected_readnoise = header['RDNOISE'+amp]
+            if rdnoise < 0.5*expected_readnoise:
+                log.error('Amp {} measured readnoise {:.2f} < 0.5 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+            elif rdnoise < 0.9*expected_readnoise:
+                log.warn('Amp {} measured readnoise {:.2f} < 0.9 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+            elif rdnoise > 2.0*expected_readnoise:
+                log.error('Amp {} measured readnoise {:.2f} > 2 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+            elif rdnoise > 1.2*expected_readnoise:
+                log.warn('Amp {} measured readnoise {:.2f} > 1.2 * expected readnoise {:.2f}'.format(
+                    amp, rdnoise, expected_readnoise))
+        else:
+            log.warn('Expected readnoise keyword {} missing'.format('RDNOISE'+amp))
+        
+        #- subtract overscan from data region and apply gain
+        jj = _parse_sec_keyword(header['DATASEC'+amp])
+        data = rawimage[jj] - overscan
+        image[kk] = data*gain
+
+    #- Load mask
+    if mask is not False and mask is not None:
+        if mask is True:
+            mask = read_mask(camera=camera, dateobs=dateobs)
+        elif isinstance(mask, (str, unicode)):
+            mask = read_mask(filename=mask)
+    else:
+        mask = np.zeros(image.shape, dtype=np.int32)
+
+    if mask.shape != image.shape:
+        raise ValueError('shape mismatch mask {} != image {}'.format(mask.shape, image.shape))
+
+    #- Divide by pixflat image
+    if pixflat is not False and pixflat is not None:
+        if pixflat is True:
+            pixflat = read_pixflat(camera=camera, dateobs=dateobs)
+        elif isinstance(pixflat, (str, unicode)):
+            pixflat = read_pixflat(filename=pixflat)
+
+        if pixflat.shape != image.shape:
+            raise ValueError('shape mismatch pixflat {} != image {}'.format(pixflat.shape, image.shape))
+
+        if np.all(pixflat != 0.0):
+            image /= pixflat
+            readnoise /= pixflat
+        else:
+            good = (pixflat != 0.0)
+            image[good] /= pixflat[good]
+            readnoise[good] /= pixflat[good]
+            mask[~good] |= ccdmask.PIXFLATZERO
+            
+        lowpixflat = (0 < pixflat) & (pixflat < 0.1)
+        if np.any(lowpixflat):
+            mask[lowpixflat] |= ccdmask.PIXFLATLOW
+
+    #- Inverse variance, estimated directly from the data (BEWARE: biased!)
+    var = image.clip(0) + readnoise**2
+    ivar = 1.0 / var
```

Comment:
I don't think it is enforced, but `readnoise` should always be > 0 such that `var` should have a minimum value of `readnoise**2` -- i.e. I don't think we are in danger of a divide-by-zero.  Note: the current line is `var = image.clip(0) + readnoise**2`, not `var = (image + readnoise**2).clip(0)` which could result in zeros if the image had large negative pixels.

I'm thinking about the statistical question about whether negative pixel values should contribute to the variance model or not.  We're in questionable statistical territory either way...